helix-view/src/document.rs

```diff
@@ -111,6 +115,144 @@ impl FromStr for Mode {
+    // the reader is determined only when a `read()` call returns `0`.
+    let buf_str = unsafe { std::str::from_utf8_unchecked_mut(&mut buf_out[..]) };
+    let mut total_read = 0usize;
+    let mut total_written = 0usize;
+    loop {
+        let (result, read, written, ..) = decoder.decode_to_str(
+            &slice[total_read..],
+            &mut buf_str[total_written..],
+            is_empty,
+        );
+
+        // These additive quantaties are needed in case the `result` was
+        // `encoding_rs::CoderResult::OutputFull` as they act as "cursors"
+        // into the input and output buffers respectively. Otherwise, it
+        // would be trapped in an endless read loop or would overwrite
+        // the output buffer.
+        total_read += read;
+        total_written += written;
+        match result {
+            encoding_rs::CoderResult::InputEmpty => {
+                debug_assert_eq!(slice.len(), total_read);
+                let read = reader.read(&mut buf)?;
+                slice = &buf[..read];
+                is_empty = read == 0;
+                total_read = 0;
+            }
+            encoding_rs::CoderResult::OutputFull => {
+                debug_assert!(slice.len() > total_read);
+                builder.append(&buf_str[..total_written]);
+                total_written = 0;
+            }
+        }
+
+        if is_empty {
+            debug_assert_eq!(reader.read(&mut buf)?, 0);
+            builder.append(&buf_str[..total_written]);
+            break;
+        }
+    }
+    let rope = builder.finish();
+    Ok((rope, encoding))
+}
+
+/// 
+pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
+    writer: &'a mut W,
+    encoding: &'static encoding_rs::Encoding,
+    rope: &'a Rope,
+) -> Result<(), Error> {
+    let mut iter = rope
+        .chunks()
+        .filter(|c| !c.is_empty())
+        .chain(std::iter::once(""));
+    let mut buf = [0u8; BUF_SIZE];
+    let mut encoder = encoding.new_encoder();
+
+    let mut chunk = iter.next().unwrap();
+    let mut is_empty = chunk.is_empty();
+    let mut total_read = 0usize;
+    let mut total_written = 0usize;
+    loop {
+        let (result, read, written, ..) =
+            encoder.encode_from_utf8(&chunk[total_read..], &mut buf[total_written..], is_empty);
+
+        total_read += read;
+        total_written += written;
+        match result {
+            encoding_rs::CoderResult::InputEmpty => {
+                debug_assert_eq!(chunk.len(), total_read);
+                debug_assert!(buf.len() >= total_written);
+                chunk = iter.next().unwrap();
+                is_empty = chunk.is_empty();
+                total_read = 0;
+            }
+            encoding_rs::CoderResult::OutputFull => {
+                debug_assert!(chunk.len() > total_read);
+                writer.write_all(&buf[..total_written]).await?;
+                total_written = 0;
+            }
+        }
```

Comment:
Same here, I believe you still need the inner nested loop, and break on `InputEmpty`.

Reply:
There are some other things that also aren't quite correct I think in the new version of these functions, but based on the commit messages it looks like they're still WIP, so I'll hold off on further comments until you're done.

Reply:
Actually, feel free to comment right now since I hope to get them merged tonight.

Reply:
See my other comment about the nested loop thing (I think the nested approach makes the code clearer).

I think the remainder of my concerns were me misunderstanding the code due to not following what it was doing properly.

Reply:
I'll mark this as resolved, since it's not a correctness issue.